### PR TITLE
Enable sliding window attention

### PR DESCRIPTION
**Pull request by abhishekkunuru6-cmyk**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Adds `spyre::sliding_window_attention`. Each block of query rows reads only the
KV slice its rows can reach and attends against that, rather than scoring the
full cache behind a band mask:

| shape | masked full attention | this PR |
|---|---|---|
| decode `Lq=1, Lkv=4096, W=64` | `[1, 4096]` | `[1, 64]` |
| prefill `Lq=Lkv=512, W=64` | `[512, 512]` | `8 x [64, 128]` |

The cache is a compact, time-ordered buffer — HF's `StaticSlidingWindowLayer`
layout. Windowing is an offset plus a length, not an `[Lq, Lkv]` mask; no ring
or modulo indexing. The buffer is `W + q_block` wide and the same width for
every block, so one allocation serves all of them: ragged blocks shift their
read rather than shrinking it.

`read_start` floors in *physical* rather than logical coordinates, which makes
it position-independent for a rolled buffer — the same value at every decode
step, so one compiled graph serves the whole generation.

**Supported:** causal windows; `key.size(2) % 64 == 0`; any `cache_seqlen`
(tokens seen) and any in-range `buffer_origin` (logical position of physical
row 0). Query length is padded up automatically; anything else raises
`Unsupported` naming the constraint. `Lk >= round_up_to_64(window_size + Lq)`
, so long sequences must be prefilled in chunks sized to the allocation.

#### Which issue(s) this PR is related to:

Related to #3073.

#### Special notes for your reviewer:

- **No unwritten-tail mask and no warmup guard in `check_window_read`, and
  that is deliberate.** A query row's coordinate is always `< cache_seqlen`, so
  causal `column <= row` already excludes every overshot column; a logical
  bound in `check_window_read` rejected most valid plans. Both become
  load-bearing only for a bidirectional window, which raises today. Rows must
  still hold **finite** values — zero-fill the allocation.

- **The `spyre_hint` nest emits no `scf.for` yet**, so the online-softmax
  accumulators are an identity today and the body computes exactly
  `softmax(scores) @ v_win`. Kept because they make the body legal once it
  tiles, and because it mirrors `spyre__sdpa_overrideable`. Tracked separately;
  the work reduction is independent of it.

- **Bidirectional is not implemented** — needs a `2W + q_block` buffer.
  `is_causal` exists because the mask op handles both; only the placement does
  not.

#### Does this PR introduce a user-facing change?

Yes. `torch.ops.spyre.sliding_window_attention(query, key, value, window_size,
is_causal, scale, cache_seqlen, buffer_origin)` under `torch.compile` on Spyre,
reading only the window from a compact KV cache rather than the whole cache
behind a mask. `cache_seqlen` and `buffer_origin` are optional and default to
the full-length exactly-full geometry.

### CODE REVIEW

**Comment by abhishekkunuru6-cmyk** [comment]
cc: @ani300 @saurabh-srivastava4 @goutham-binnadi-gopala

**Comment by ani300** [comment]
Heads-up: this will break on rebase/merge onto current `main`. Upstream #3811 (merged 2026-08-19) removed `spyre::copy_f`, splitting it into `copy_forced` (mutating) and `opaque_copy_` (functional at trace time). `_windowed_attention` in `decompositions.py` is the last caller of `copy_f` in the tree, so once this PR sees post-#3811 `main` it'll fail with `AttributeError` from the spyre op namespace.

`opaque_copy_` is the form #3811 prescribes for code inside a `torch.compile`-traced decomposition (AOTAutograd functionalization rejects the mutating variant). Same `(value, acc) -> Tensor` convention, so it's a pure rename. Diff to fold into this branch:

```diff
@@ -659,6 +659,12 @@ def _windowed_attention(
         )
         q_rows = query[:, :, q_start:q_end, :]
 
+        # opaque_copy_, not copy_f: upstream #3811 removed spyre::copy_f and
+        # split it into copy_forced (mutating) and opaque_copy_ (functional at
+        # trace time). This runs inside a decomposition traced by torch.compile,
+        # where AOTAutograd functionalization rejects the mutating form, so
+        # opaque_copy_ is the prescribed replacement — same (value, acc) call
+        # convention, so this is a rename.
         # SDPA's "sparse via reduction" construction. On a single pass M is
         # -inf, so correction is exp(-inf) == 0 and the running terms drop out.
         m_reduced = torch.full(
@@ -700,19 +706,21 @@ def _windowed_attention(
                         exp_scores = torch.exp(scores - max_running.unsqueeze(-1))
                         correction = torch.exp(running_max - max_running)
 
-                        denominator = torch.ops.spyre.copy_f(
+                        denominator = torch.ops.spyre.opaque_copy_(
                             denominator * correction + exp_scores.sum(dim=-1),
                             denominator,
                         )
-                        output = torch.ops.spyre.copy_f(
+                        output = torch.ops.spyre.opaque_copy_(
                             output * correction.unsqueeze(-1)
                             + torch.matmul(exp_scores, v_win),
                             output,
                         )
-                        running_max = torch.ops.spyre.copy_f(max_running, running_max)
+                        running_max = torch.ops.spyre.opaque_copy_(
+                            max_running, running_max
+                        )
 
         out_blocks.append(
-            torch.ops.spyre.copy_f(output / denominator.unsqueeze(-1), output)
+            torch.ops.spyre.opaque_copy_(output / denominator.unsqueeze(-1), output)
         )
 
     return torch.cat(out_blocks, dim=2)
```

(Verified against my SWA integration branch, where this migration is already applied.)

**Comment by ani300** [comment]
Offering an add-on for the op here rather than as a competing PR, since it stacks directly on `swa-window-roll`. It came out of wiring `spyre::sliding_window_attention` into the hf-adapters Gemma 3/4 path, where two gaps surfaced:

**1. `valid_start` — left-padding exclusion.** An offset-and-length window cannot skip padded columns, so left-padding has to travel to the op explicitly. Without it the op silently attends the left-pad K/V columns: the logit magnitude barely moves but the argmax flips, which reads like a subtle numeric bug and is actually a missing input. This adds an optional per-batch `valid_start` argument (a first-attendable column per sequence) that masks those columns out, with a fast path that is numerically identical to passing nothing when every entry is 0.

**2. Gemma 4 geometry coverage.** Two tests for shapes the hf-adapters integration actually calls: `head_dim=256` GQA prefill (16 q-heads from 8 kv-heads — four sticks per row, where the rest of the file uses one or two), and the anchored decode step (`seqlen_q=1` against an exactly-full 1088-row compact buffer).

Three commits, split so the coverage lands independently of the feature:

- `test(swa)` — head_dim-256 GQA + anchored-decode coverage
- `feat(swa)` — torch-free `valid_start` helpers for the window band
- `feat(swa)` — `valid_start` wired through the op

**Verification** (rebased onto the current tip of `swa-window-roll`): `tests/inductor/test_sliding_window_attention.py` + `tests/inductor/test_kv_window.py` → **131 passed**. The only two failures (`test_query_length_not_a_multiple_of_the_block`, `test_ragged_query_and_window_together`) reproduce identically on unmodified `swa-window-roll` (same 30749/51200 mismatch) — pre-existing, not touched by this change. `pre-commit run` clean on all five files.

Apply with `git am` (or cherry-pick the branch if you'd rather — happy to push it wherever is convenient). Patches below:

<details>
<summary>Three patches (<code>git am</code>-ready)</summary>

```patch
From 016d88c31998ceeed46fc52387d51948423ffa42 Mon Sep 17 00:00:00 2001
From: Antoni Viros <aviros@ibm.com>
Date: Tue, 18 Aug 2026 21:11:58 +0000
Subject: [PATCH 1/3] test(swa): cover Gemma 4's head_dim 256 and anchored
 decode

Signed-off-by: Antoni Viros <aviros@ibm.com>
---
 .../inductor/test_sliding_window_attention.py | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/tests/inductor/test_sliding_window_attention.py b/tests/inductor/test_sliding_window_attention.py
index f0b104c1..214f7da4 100644
--- a/tests/inductor/test_sliding_window_attention.py
+++ b/tests/inductor/test_sliding_window_attention.py
@@ -254,6 +254,13 @@ class TestSlidingWindowAttention(unittest.TestCase):
         query, key, value = _inputs(1, 8, 2, 100, 512)
         compare_with_cpu(_attention, query, key, value, 100, run_eager=False)
 
+    def test_prefill_head_dim_256_gqa(self):
+        # Gemma 4's sliding layers: 16 query heads from 8 KV heads, head_dim 256,
+        # W=1024. head_dim 256 is four sticks per row where the rest of this file
+        # uses one or two, and kv_window hands back a transposed slice.
+        query, key, value = _inputs(1, 16, 8, 512, 512, head_dim=256)
+        compare_with_cpu(_attention, query, key, value, 1024, run_eager=False)
+
 
 class TestCompactCache(unittest.TestCase):
     """cache_seqlen != key.size(2): a compact cache, rolled or still filling.
@@ -354,6 +361,19 @@ class TestCompactCache(unittest.TestCase):
             _rolled_attention, query, key, value, window, cache_seqlen, run_eager=False
         )
 
+    def test_anchored_decode_gemma4(self):
+        # The exact geometry hf-adapters will call every decode step: a 1088-row
+        # compact buffer declared exactly full, and a single query row
+        # (seqlen_q=1) fed straight to the op. The anchored design dropped the
+        # 64-row query stick, so the real decode shape is (1, 16, 1, 256), with
+        # head_dim 256 and W=1024. This is the one shape that must be right for
+        # the integration to work at all.
+        key, value = _compact_kv(1, 8, 1088, 1088, head_dim=256)
+        query = cached_randn((1, 16, 1, 256), differentiation=1, dtype=torch.float16)
+        compare_with_cpu(
+            _rolled_attention, query, key, value, 1024, 1088, run_eager=False
+        )
+
 
 if __name__ == "__main__":
     unittest.main()
-- 
2.52.0

From 9292a2120f27e64bf9d9cb1d2dcb35fd1212adf0 Mon Sep 17 00:00:00 2001
From: Antoni Viros <aviros@ibm.com>
Date: Tue, 18 Aug 2026 21:19:35 +0000
Subject: [PATCH 2/3] feat(swa): torch-free valid_start helpers for the window
 band

Signed-off-by: Antoni Viros <aviros@ibm.com>
---
 tests/inductor/test_kv_window.py             | 35 ++++++++++++++
 torch_spyre/_inductor/sliding_window_plan.py | 48 ++++++++++++++++++++
 2 files changed, 83 insertions(+)

diff --git a/tests/inductor/test_kv_window.py b/tests/inductor/test_kv_window.py
index a4637c55..5b2e717e 100644
--- a/tests/inductor/test_kv_window.py
+++ b/tests/inductor/test_kv_window.py
@@ -45,6 +45,9 @@ import torch
 from torch_spyre._inductor.sliding_window_plan import (
     STICK,
     SlidingWindowPlan,
+    band_batch,
+    band_valid_start,
+    check_valid_start,
     check_window_read,
     default_buffer_origin,
     plan_sliding_window,
@@ -916,5 +919,37 @@ class TestKVWindowOp:
         compare_with_cpu(fn, cache, cache, run_eager=False)
 
 
+class TestValidStart:
+    """The three torch-free helpers behind the valid_start band."""
+
+    def test_none_and_all_zero_mask_nothing(self):
+        # A caller with no padding must not pay for a per-batch band.
+        assert band_valid_start(None) is None
+        assert band_valid_start([]) is None
+        assert band_valid_start([0, 0]) is None
+        assert band_batch([0, 0]) == 1
+
+    def test_uniform_nonzero_stays_broadcast(self):
+        # Same threshold for every sequence: one row of band, broadcast over batch.
+        assert band_valid_start([40, 40]) == [40, 40]
+        assert band_batch([40, 40]) == 1
+
+    def test_ragged_widens_to_batch(self):
+        assert band_batch([0, 40]) == 2
+        assert band_batch([40, 40, 7]) == 3
+
+    def test_check_rejects_wrong_length(self):
+        assert "2 entries" in check_valid_start([0, 40], 3, 1088)
+
+    def test_check_rejects_out_of_range(self):
+        assert "outside" in check_valid_start([-1], 1, 1088)
+        assert "outside" in check_valid_start([1089], 1, 1088)
+
+    def test_check_accepts_valid(self):
+        assert check_valid_start(None, 4, 1088) is None
+        assert check_valid_start([0, 40], 2, 1088) is None
+        assert check_valid_start([1088], 1, 1088) is None
+
+
 if __name__ == "__main__":
     pytest.main([__file__, "-v"])
diff --git a/torch_spyre/_inductor/sliding_window_plan.py b/torch_spyre/_inductor/sliding_window_plan.py
index 7bb3a1d1..ea90b962 100644
--- a/torch_spyre/_inductor/sliding_window_plan.py
+++ b/torch_spyre/_inductor/sliding_window_plan.py
@@ -412,3 +412,51 @@ def plan_sliding_window(
         cache_capacity=cache_capacity,
         buffer_origin=buffer_origin,
     )
+
+
+def band_valid_start(valid_start: list[int] | None) -> list[int] | None:
+    """The ``valid_start`` the band must actually apply, or None when it masks nothing.
+
+    A caller with no left padding passes zeros rather than tracking whether it has
+    any; collapsing that to None here is what keeps the band broadcast over batch
+    and keeps ``block_is_fully_attended``'s skip available.
+    """
+    if not valid_start or max(valid_start) <= 0:
+        return None
+    return list(valid_start)
+
+
+def band_batch(valid_start: list[int] | None) -> int:
+    """Leading dimension of the band: 1 unless the threshold differs per sequence.
+
+    The band is ``q_block x buffer_width`` per distinct threshold, so a uniform one
+    stays a single broadcast row rather than one copy per batch entry.
+    """
+    effective = band_valid_start(valid_start)
+    if effective is None or min(effective) == max(effective):
+        return 1
+    return len(effective)
+
+
+def check_valid_start(
+    valid_start: list[int] | None, batch: int, cache_seqlen: int
+) -> str | None:
+    """Why this ``valid_start`` is invalid, or None.
+
+    Strings not exceptions, matching ``check_window_read`` -- this module stays
+    free of torch and of the backend's error classes. Batch is a tensor property,
+    so this cannot live in ``rejection_reason``, which answers placement questions
+    from integers alone.
+    """
+    if valid_start is None:
+        return None
+    if len(valid_start) != batch:
+        return (
+            f"valid_start has {len(valid_start)} entries for a batch of {batch}"
+        )
+    for entry in valid_start:
+        if entry < 0 or entry > cache_seqlen:
+            return (
+                f"valid_start={entry} outside [0, cache_seqlen={cache_seqlen}]"
+            )
+    return None
-- 
2.52.0

From 9604060866e95c7bb7b63e0c0ab3ce2ad0657d79 Mon Sep 17 00:00:00 2001
From: Antoni Viros <aviros@ibm.com>
Date: Tue, 18 Aug 2026 21:35:30 +0000
Subject: [PATCH 3/3] feat(swa): valid_start to exclude left-pad columns from
 the window

Signed-off-by: Antoni Viros <aviros@ibm.com>
---
 .../inductor/test_sliding_window_attention.py | 61 +++++++++++++++++++
 torch_spyre/_inductor/customops.py            | 40 ++++++++++--
 torch_spyre/_inductor/decompositions.py       | 20 +++++-
 torch_spyre/_inductor/sliding_window_plan.py  |  8 +--
 4 files changed, 116 insertions(+), 13 deletions(-)

diff --git a/tests/inductor/test_sliding_window_attention.py b/tests/inductor/test_sliding_window_attention.py
index 214f7da4..d1c1284e 100644
--- a/tests/inductor/test_sliding_window_attention.py
+++ b/tests/inductor/test_sliding_window_attention.py
@@ -147,6 +147,36 @@ def _rolled_reference(query, key, value, window_size, cache_seqlen, buffer_origi
     )
 
 
+def _reference_with_valid_start(query, key, value, window_size, valid_start):
+    """``_rolled_reference`` for an exactly-full buffer, additionally excluding
+    physical rows below ``valid_start`` -- the left-padding an offset-and-length
+    window cannot express.
+
+    One threshold per batch entry, so ``valid_start`` is a list even when uniform.
+    """
+    seqlen_q, capacity = query.size(2), key.size(2)
+    rows = torch.arange(seqlen_q) + (capacity - seqlen_q)
+    columns = torch.arange(capacity)
+    delta = rows.unsqueeze(-1) - columns.unsqueeze(0)
+    allowed = (delta >= 0) & (delta < window_size)
+    starts = torch.tensor(valid_start).view(-1, 1, 1)
+    allowed = allowed.unsqueeze(0) & (columns.view(1, 1, -1) >= starts)
+    mask = torch.zeros(allowed.shape, dtype=query.dtype)
+    mask.masked_fill_(~allowed, float("-inf"))
+    return F.scaled_dot_product_attention(
+        query, key, value, attn_mask=mask.unsqueeze(1)
+    )
+
+
+def _valid_start_attention(q, k, v, window_size, valid_start):
+    """The op with an explicit valid_start on spyre, the reference on CPU."""
+    if q.device.type == "spyre":
+        return torch.ops.spyre.sliding_window_attention(
+            q, k, v, window_size, True, None, k.size(2), 0, valid_start
+        )
+    return _reference_with_valid_start(q, k, v, window_size, valid_start)
+
+
 def _rolled_attention(q, k, v, window_size, cache_seqlen, buffer_origin=None):
     """Dispatch for a compact cache: the op on spyre with an explicit
     ``cache_seqlen`` (it cannot default to ``k.size(2)`` here -- that IS the
@@ -374,6 +404,37 @@ class TestCompactCache(unittest.TestCase):
             _rolled_attention, query, key, value, 1024, 1088, run_eager=False
         )
 
+    def test_valid_start_excludes_padded_columns(self):
+        # 17 rows of left padding inside the window: without valid_start they are
+        # attended, and the reference proves the difference is visible.
+        key, value = _compact_kv(1, 8, 1088, 1088)
+        query = cached_randn((1, 8, 64, 64), differentiation=1, dtype=torch.float16)
+        compare_with_cpu(
+            _valid_start_attention, query, key, value, 1024, [17], run_eager=False
+        )
+
+    def test_valid_start_per_sequence(self):
+        # Ragged batch: the band widens to [B, 1, q, W'] only for this case.
+        key, value = _compact_kv(2, 8, 1088, 1088)
+        query = cached_randn((2, 8, 64, 64), differentiation=1, dtype=torch.float16)
+        compare_with_cpu(
+            _valid_start_attention, query, key, value, 1024, [0, 40], run_eager=False
+        )
+
+    def test_all_zero_valid_start_matches_no_valid_start(self):
+        # The fast path must be numerically identical to passing nothing.
+        key, value = _compact_kv(1, 8, 1088, 1088)
+        query = cached_randn((1, 8, 64, 64), differentiation=1, dtype=torch.float16)
+
+        def attention(q, k, v, window_size):
+            if q.device.type == "spyre":
+                return torch.ops.spyre.sliding_window_attention(
+                    q, k, v, window_size, True, None, k.size(2), 0, [0]
+                )
+            return _rolled_reference(q, k, v, window_size, k.size(2))
+
+        compare_with_cpu(attention, query, key, value, 1024, run_eager=False)
+
 
 if __name__ == "__main__":
     unittest.main()
diff --git a/torch_spyre/_inductor/customops.py b/torch_spyre/_inductor/customops.py
index 8cbd3aba..2dc5f0f3 100644
--- a/torch_spyre/_inductor/customops.py
+++ b/torch_spyre/_inductor/customops.py
@@ -20,6 +20,7 @@ from torch_spyre.ops.eager import compile_once
 from torch_spyre.ops.fallbacks import warn_fallback
 
 from .errors import Unsupported
+from .sliding_window_plan import band_batch, band_valid_start
 
 aten = torch.ops.aten
 
@@ -876,6 +877,7 @@ def sliding_window_attention(  # type: ignore[empty-body]
     scale: Optional[float] = None,
     cache_seqlen: Optional[int] = None,
     buffer_origin: Optional[int] = None,
+    valid_start: Optional[list[int]] = None,
 ) -> torch.Tensor:
     """
     Sliding-window attention entry point.
@@ -894,6 +896,15 @@ def sliding_window_attention(  # type: ignore[empty-body]
     MUST pass its true origin, or every read lands past the data, in bounds
     and unmasked. See ``SlidingWindowPlan`` for both.
 
+    ``valid_start`` is one **logical** column coordinate per batch entry;
+    columns strictly below it are never attended, whatever the window says. It
+    exists for left-padded prompts, whose pad columns sit inside the window and
+    which an offset-and-length window cannot otherwise exclude. ``None`` or
+    all-zero costs nothing; a uniform threshold keeps the band broadcast over
+    batch; only a ragged one widens it. Coordinates are logical, matching
+    ``cache_seqlen``, so a caller passing ``buffer_origin=0`` passes physical
+    row indices.
+
     Allocation contract: ``Lk >= round_up_to_64(window_size + Lq - 1)`` for
     the ``Lq`` of this call, so prefill long sequences in chunks;
     ``rejection_reason`` names the required row count when ``Lk`` is short.
@@ -918,6 +929,7 @@ def _(
     scale: Optional[float] = None,
     cache_seqlen: Optional[int] = None,
     buffer_origin: Optional[int] = None,
+    valid_start: Optional[list[int]] = None,
 ) -> torch.Tensor:
     return query.new_empty(query.size())
 
@@ -934,13 +946,16 @@ def window_band_mask(
     is_causal: bool,
     dtype: torch.dtype,
     device: torch.device,
+    valid_start: Optional[list[int]] = None,
 ) -> torch.Tensor:
     """
     Additive band over one Q block's KV window.
 
-    Shape: [1, 1, q_block, buffer_width]; 0.0 = keep, -inf = masked. Both
-    leading axes are size 1 and broadcast over batch and heads, which keeps
-    this at q_block x buffer_width elements rather than one copy per head.
+    Shape: [1, 1, q_block, buffer_width] normally; the leading axis is 1
+    unless ``valid_start`` differs across the batch, in which case it is B.
+    0.0 = keep, -inf = masked. The head axis stays size 1 and broadcasts,
+    which keeps this at q_block x buffer_width elements per batch row rather
+    than one copy per head.
 
     Query row ``q_row_origin + i`` (an absolute KV-cache coordinate: row 0 of
     this block sits at ``seqlen_kv - seqlen_q + q_block*n``) may attend column
@@ -961,6 +976,9 @@ def window_band_mask(
     ``column < cache_seqlen``. Only a bidirectional window, which raises
     today, would need one.
 
+    ``valid_start``, if given, additionally excludes columns strictly below
+    its per-batch-entry threshold -- see ``sliding_window_attention``.
+
     Built entirely on CPU so the in-place ops stay opaque to torch.compile,
     matching spyre.causal_mask's rationale.
     """
@@ -971,9 +989,18 @@ def window_band_mask(
         allowed = (delta >= 0) & (delta < window_size)
     else:
         allowed = delta.abs() < window_size
+    effective = band_valid_start(valid_start)
+    if effective is None:
+        allowed = allowed.unsqueeze(0)
+    elif min(effective) == max(effective):
+        # Uniform threshold: still one broadcast row, not one per sequence.
+        allowed = (allowed & (column.unsqueeze(0) >= effective[0])).unsqueeze(0)
+    else:
+        starts = torch.tensor(effective, device="cpu").view(-1, 1, 1)
+        allowed = allowed.unsqueeze(0) & (column.view(1, 1, -1) >= starts)
     mask_cpu = torch.zeros(allowed.shape, dtype=dtype, device="cpu")
     mask_cpu.masked_fill_(~allowed, float("-inf"))
-    return mask_cpu.unsqueeze(0).unsqueeze(0).to(device=device)
+    return mask_cpu.unsqueeze(1).to(device=device)
 
 
 @window_band_mask.register_fake
@@ -986,8 +1013,11 @@ def _(
     is_causal: bool,
     dtype: torch.dtype,
     device: torch.device,
+    valid_start: Optional[list[int]] = None,
 ) -> torch.Tensor:
-    return torch.empty(1, 1, q_block, buffer_width, dtype=dtype, device=device)
+    return torch.empty(
+        band_batch(valid_start), 1, q_block, buffer_width, dtype=dtype, device=device
+    )
 
 
 @torch.library.custom_op("spyre::kv_window", mutates_args=(), device_types="spyre")
diff --git a/torch_spyre/_inductor/decompositions.py b/torch_spyre/_inductor/decompositions.py
index 51435dd7..9e021c8f 100644
--- a/torch_spyre/_inductor/decompositions.py
+++ b/torch_spyre/_inductor/decompositions.py
@@ -36,6 +36,8 @@ from .constants import DEVICE_NAME, FP8_E4M3FN_MAX, FP8_E4M3FN_MIN
 from .errors import Unsupported
 from .sliding_window_plan import (
     SlidingWindowPlan,
+    band_valid_start,
+    check_valid_start,
     check_window_read,
     plan_sliding_window,
     query_blocking,
@@ -602,6 +604,7 @@ def _windowed_attention(
     plan: SlidingWindowPlan,
     scaling_factor: float,
     num_heads: int,
+    valid_start: list[int] | None = None,
 ) -> torch.Tensor:
     """spyre__sdpa_overrideable's body, tiled over buffer_width not max_seqlen_kv.
 
@@ -628,7 +631,12 @@ def _windowed_attention(
         k_win, v_win = torch.ops.spyre.kv_window(
             key, value, read_start, buffer_width, num_heads
         )
-        fully_attended = plan.block_is_fully_attended(block_index)
+        # A valid_start that masks anything makes the band load-bearing even for a
+        # block the window alone fully covers.
+        fully_attended = (
+            plan.block_is_fully_attended(block_index)
+            and band_valid_start(valid_start) is None
+        )
         band = (
             None
             if fully_attended
@@ -646,6 +654,7 @@ def _windowed_attention(
                 plan.is_causal,
                 query.dtype,
                 query.device,
+                valid_start,
             )
         )
         q_rows = query[:, :, q_start:q_end, :]
@@ -719,6 +728,7 @@ def spyre_sliding_window_attention(
     scale: float | None = None,
     cache_seqlen: int | None = None,
     buffer_origin: int | None = None,
+    valid_start: list[int] | None = None,
 ) -> torch.Tensor:
     """Sliding-window attention: each Q block attends only its own KV slice.
 
@@ -746,6 +756,10 @@ def spyre_sliding_window_attention(
     if cache_seqlen is None:
         cache_seqlen = cache_capacity
 
+    reason = check_valid_start(valid_start, batch_size, cache_seqlen)
+    if reason is not None:
+        raise Unsupported(f"sliding_window_attention: {reason}")
+
     if scale is not None and scale < 0:
         # math.sqrt would otherwise raise a bare ValueError.
         raise Unsupported(
@@ -802,7 +816,9 @@ def spyre_sliding_window_attention(
             dim=2,
         )
 
-    output = _windowed_attention(query, key, value, plan, scaling_factor, num_heads)
+    output = _windowed_attention(
+        query, key, value, plan, scaling_factor, num_heads, valid_start
+    )
     return output[:, :, pad_rows:, :] if pad_rows else output
 
 
diff --git a/torch_spyre/_inductor/sliding_window_plan.py b/torch_spyre/_inductor/sliding_window_plan.py
index ea90b962..e53c79cb 100644
--- a/torch_spyre/_inductor/sliding_window_plan.py
+++ b/torch_spyre/_inductor/sliding_window_plan.py
@@ -451,12 +451,8 @@ def check_valid_start(
     if valid_start is None:
         return None
     if len(valid_start) != batch:
-        return (
-            f"valid_start has {len(valid_start)} entries for a batch of {batch}"
-        )
+        return f"valid_start has {len(valid_start)} entries for a batch of {batch}"
     for entry in valid_start:
         if entry < 0 or entry > cache_seqlen:
-            return (
-                f"valid_start={entry} outside [0, cache_seqlen={cache_seqlen}]"
-            )
+            return f"valid_start={entry} outside [0, cache_seqlen={cache_seqlen}]"
     return None
-- 
2.52.0


```

</details>

**Comment by dgrove-oss** [comment]
#4140 made another adjustment to the `copy_forced` API.  Details are in the PR, but the TL'DR for this PR is that I think you just replace `opqaue_copy_` by `copy_forced`

**Comment by abhishekkunuru6-cmyk** [comment]
@dgrove-oss @ani300 
Done — all four `_windowed_attention` sites now use `copy_forced`
(`decompositions.py:702,706,711,716`). Straight rename: same
`MutationLayoutSHOULDREMOVE` lowering, same reassignment contract, and src/dst shapes
match at every site.

| Configuration | SWA suite |
|---|---|
| `HBM_POOL_PLANNING=0` | 26 passed |
| default (`=1`) | 14 failed / 12 passed |
| default, with #4193 | 26 passed |

The 14 failures aren't SWA's — they're a pre-existing hole in `hbm_pool_planning.py`,
filed as #4193. A pool offset is written into the `layout.allocation` dict
that co-allocated names share, so it relocates names that were never pool candidates.
Each block's `copy_forced(output / denominator, output)` result is pooled in the
attention bundle and read by the `cat` bundle, which then addresses it pool-relative
while owning no pool. Multi-block prefill only; decode has no second bundle to escape
into.

**This PR stays red in CI until that one lands** (`HBM_POOL_PLANNING` defaults on).

**Comment by dgrove-oss** [comment]
Have you verified your claim of a pool allocation bug against current main?

**Comment by dgrove-oss** [comment]
> Have you verified your claim of a pool allocation bug against current main?

Never mind.  I did as part of reviewing the bug fix PR.

**Review by dgrove-oss — APPROVED** [review]
Summary

Adds spyre::sliding_window_attention, spyre::kv_window, and spyre::window_band_mask, plus placement arithmetic in sliding_window_plan.py. Each Q block reads a fixed-width KV slice (window_size + q_block - 1 rows) instead of scoring the full cache behind a band mask, so cost scales with the window rather than cache length. Supports prefill, decode, GQA, chunked prefill, and compact/rolled caches via explicit cache_seqlen/buffer_origin.

Review

BLOCKERs

None found.

SUGGESTIONs

- [ ] The online-softmax accumulator scaffolding in _windowed_attention (m_reduced, denominator_reduced, running-max/correction machinery) is dead code today since buffer_width is never tiled — it always executes exactly one iteration where correction = exp(-inf) = 0. The PR body acknowledges this ("kept because they make the body legal once it tiles"). Reasonable to keep given it mirrors spyre__sdpa_overrideable, but worth flagging so a future reader doesn't mistake it for live multi-block softmax logic.
- [ ] No docs/ update for the three new ops, though this matches the existing convention of not documenting internal spyre::* customops (e.g. spyre__sdpa_overrideable isn't documented either) — not a gap introduced by this PR specifically.

QUESTIONs

- [ ] kv_window's docstring states the row-order/contiguity precondition for a compact buffer is "unproven on this backend" since aten::roll has no lowering. Is there a follow-up tracked for whoever integrates this (e.g. HF adapters) to actually implement/validate the roll, or does the caller currently guarantee order some other way?
- [ ] Bidirectional windows (is_causal=False) raise Unsupported and are called out as a known gap needing a 2W + q_block buffer — confirming this is intentionally out of scope for this PR and tracked separately (per #3073)?

Notes

- License headers, import regex convention, and DCO sign-off all correct.
- pre-commit run --all-files on the changed files passes clean (ruff, mypy, yamlfmt, import-regex hook).
- All CI checks pass, including the new Test Sliding Window Attention and Test Kv Window suites.
- Test coverage is exceptional: three-tier structure (pure-integer placement algebra → CPU algorithm-equivalence proofs → on-device op tests), including regression tests for specific bugs found during development (collapsed read-start bug, non-guarded reach check, etc.), edge cases (ragged query padding, non-stick-aligned windows/cache_seqlen, GQA, head_dim 128/256, valid_start left-padding), and an explicit combinatorial "no leakage/no coverage gap" sweep.
- check_window_read/rejection_reason give named, actionable error messages via Unsupported, per project convention.
- GQA expansion in spyre_kv_window is consistent between fake and real lowering and mirrors the existing spyre__sdpa_overrideable pattern.

Verdict

APPROVE — high-quality implementation with thorough test coverage and clear documentation of design tradeoffs. The suggestions above are non-blocking; the questions are for the author's context, not defects.

**Comment by dgrove-oss** [comment]
@abhishekkunuru6-cmyk please fix the commit without a signature and then we should be in good shape to merge.

**Comment by abhishekkunuru6-cmyk** [comment]
@dgrove-oss Fixed ,  all 12 commits now Verified.
